File: 3501-delete-nodes-from-linked-list-present-in-array/delete-nodes-from-linked-list-present-in-array.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* modifiedList(vector<int>& nums, ListNode* head) {
        ListNode dummy(0, head);
        unordered_set<int> numsSet{nums.begin(), nums.end()};

        for (ListNode* curr = &dummy; curr->next != nullptr;)
            if (numsSet.contains(curr->next->val))
                curr->next = curr->next->next;
            else
                curr = curr->next;

        return dummy.next;
    }
};

File: 3501-delete-nodes-from-linked-list-present-in-array/delete-nodes-from-linked-list-present-in-array.java
class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        int max = -1;
        for (int num : nums)
            max = Math.max(max, num);
        boolean[] freq = new boolean[max + 1];

        for (int num : nums)
            freq[num] = true;

        ListNode temp = new ListNode();
        ListNode curr = temp;

        while (head != null) {
            if (head.val >= freq.length || freq[head.val] == false) {
                curr.next = head;
                curr = curr.next;
            }

            head = head.next;
        }

        curr.next = null;
        return temp.next;
    }
}